### PR TITLE
[persist] Generalize some spine internals

### PR DESCRIPTION
**Pull request by bkirwi**
This changes Persist's spine implementation from a binary tree to an N-ary tree where N=2.

Obviously this should have no actual impact on its behaviour -- but it sets us up to vary N more easily in the future.

### Motivation

#16607 

### Tips for reviewer

I pulled in `arrayvec` here - a new dep, but I think it's a good one, and it's tailor-made for this sort of case.

I think this is correct for N=2; while I've made some effort to make it more likely to work for N>2, I haven't tested it, and would prefer to take that as a followup. (Though if you notice anything obviously wrong here I'm happy to shake that out early of course!)

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [inline src/persist-client/src/internal/trace.rs:1441]: This isn't what we'd want in the general case, since when N > 2 a batch could have more than one batch but not be merging yet. I think the "right" way to implement this (given existing usage) is as a merge of all batches in the layer, even when there is no already-active merge.

**Comment by bkirwi** [comment]
Local benchmarks show this performance-neutral, or maybe a slight improvement, at least on my box. Running a subset of nightlies including some benchmarking here: https://buildkite.com/materialize/nightly/builds/7321

I did a quick check locally setting `BATCHES_PER_LEVEL = 4`; the only unit tests that failed seemed to be ones that explicitly depended on the exact sets of batches being merged, which is heartening, but I haven't done more extensive testing.

**Comment by bkirwi** [comment]
@danhhz - curious what you think of the approach before I get this ready to merge! (I don't think we've discussed how we want to go from 2 to N, but it seems like we'd need to drop the fixed enum in any case, and this feels like the way to do it that sticks closest to the existing structure...)

**Review by danhhz — COMMENTED** [review]
`arrayvec` is nice, I don't mind the dep.

I think the question this immediately raises for me is how dynamic we want these knobs to be. `const BATCHES_PER_LEVEL` fixes it for all environments running a certain version of code, whereas I think there's probably some options (`smallvec`?) that let us hook it up to a dyncfg/per-shard-override/etc. I confess I've been picturing something more like the latter, but my mind could certainly be changed! Maybe this larger question is something to hash out at the next persist sync? That said, I don't want your compaction 2.0 pipeline to stall and it seems pretty mechanical to start with `arrayvec` and swap something else in later.

Overall, very nice work on the generalization here! I've only scanned it in places, but most of the diffs are pretty obviously behavior-neutral

**Comment by bkirwi** [comment]
> I think the question this immediately raises for me is how dynamic we want these knobs to be.

Yeah, totally. My focus here was on swapping out the structure without changing the runtime characteristics by eg. adding an extra layer of indirection. That said: swapping `ArrayVec` out for `Vec` doesn't show up in the `trace` benchmarks, so maybe it's fine? Generally happy to go with whatever seems like the most boring choice... should be easy to swap later.

As far as an end state: definitely agree it would be nice to change this dynamically. If we think that we're likely to only have relatively small values of N, it seems like we could probably still get away with the `ArrayVec` with some maximum size... but if our maximum size is large we probably need to eat the extra allocation. I'm not blocked on this yet, so happy to discuss live if that's easier!

**Comment by danhhz** [comment]
Ah crud. I'd meant for us to hash this out in the persist sync or our 1:1 this week, but forgot!

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:1350]: Added validations that full batches have a corresponding fueling merge, and that the id of the merge matches the overall id of the layer.

File: src/persist-client/src/internal/trace.rs
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:1388]: I experimented with this, but since this is the only place we could use the enum, it it ended up making the diff less clear overall. Will definitely keep this in mind as we make more changes though!

**Comment by bkirwi** [comment]
From some off-PR discussion, the current idea is to go ahead with the arrayvec (since it has the most similar perf profile to the old tuples) for now; and with the understanding that we may swap this out in the fulness of time.

Lots of red in the nightlies, but going through them it seems like everything is a known issue... opening for review.

File: src/persist-client/src/internal/trace.rs
Comment (danhhz) [reply src/persist-client/src/internal/trace.rs:945]: Defer to you on this one!

**Review by danhhz — APPROVED** [review]
\o/

**Review by ParkMyCar — APPROVED** [review]
Took a quick pass and it looks good to me! Still haven't fully groked how spine works though

**Comment by bkirwi** [comment]
This seems to trigger sporadic failures in the output consistency test that aren't present on `main`. Since this is a pretty sensitive bit of code, I'll try and get to the bottom of that before merging anything in.

**Comment by bkirwi** [comment]
Checked, and the Postgres consistency failures have the same result on main - convenient that they have the repro steps in the log! The higher failure rate on this branch seems like they a coincidence after all...

**Pull request merged**
